pkg/app/api/grpcapi/web_api.go

```diff
@@ -1350,52 +1344,159 @@ func (a *WebAPI) getInsightDataForDeployFrequency(ctx context.Context, projectID
+			Value:    projectID,
+		},
+		{
+			Field:    "CreatedAt",
+			Operator: ">=",
+			Value:    targetRangeFrom.Unix(),
+		},
+		{
+			Field:    "CreatedAt",
+			Operator: "<",
+			Value:    targetRangeTo.Unix(), // target's finish time on unix time
+		},
+	}
+
+	if applicationID != "" {
+		filters = append(filters, datastore.ListFilter{
+			Field:    "ApplicationId",
+			Operator: "==",
+			Value:    applicationID,
+		})
+	}
+
+	pageSize := 50
+	deployments, err := a.deploymentStore.ListDeployments(ctx, datastore.ListOptions{
+		PageSize: pageSize,
+		Filters:  filters,
+	})
+	if err != nil {
+		a.logger.Error("failed to get deployments", zap.Error(err))
+		return nil, status.Error(codes.Internal, "Failed to get deployments")
+	}
+
+	return &model.InsightDataPoint{
+		Timestamp: targetRangeFrom.Unix(),
+		Value:     float32(len(deployments)),
+	}, nil
+}
+
+// getInsightDataForChangeFailureRate accumulate insight data in target range for change failure rate
+// This function is temporary implementation for front end.
+func (a *WebAPI) getInsightDataForChangeFailureRate(
+	ctx context.Context,
+	projectID string,
+	applicationID string,
+	targetRangeFrom time.Time,
+	targetRangeTo time.Time) (*model.InsightDataPoint, error) {
+
+	commonFilters := []datastore.ListFilter{
+		{
+			Field:    "ProjectId",
+			Operator: "==",
+			Value:    projectID,
+		},
+		{
+			Field:    "CreatedAt",
+			Operator: ">=",
+			Value:    targetRangeFrom.Unix(),
+		},
+		{
+			Field:    "CreatedAt",
+			Operator: "<",
+			Value:    targetRangeTo.Unix(), // target's finish time on unix time
+		},
+	}
+
+	if applicationID != "" {
+		commonFilters = append(commonFilters, datastore.ListFilter{
+			Field:    "ApplicationId",
+			Operator: "==",
+			Value:    applicationID,
+		})
+	}
+
+	filterForSuccessDeploy := []datastore.ListFilter{
+		{
+			Field:    "Status",
+			Operator: "==",
+			Value:    model.DeploymentStatus_DEPLOYMENT_SUCCESS,
+		},
+	}
+
+	filterForFailureDeploy := []datastore.ListFilter{
+		{
+			Field:    "Status",
+			Operator: "==",
+			Value:    model.DeploymentStatus_DEPLOYMENT_FAILURE,
+		},
+	}
+
+	pageSize := 50
+	successDeployments, err := a.deploymentStore.ListDeployments(ctx, datastore.ListOptions{
+		PageSize: pageSize,
+		Filters:  append(filterForSuccessDeploy, commonFilters...),
+	})
+	if err != nil {
+		a.logger.Error("failed to get deployments", zap.Error(err))
+		return nil, status.Error(codes.Internal, "Failed to get deployments")
+	}
+
+	failureDeployments, err := a.deploymentStore.ListDeployments(ctx, datastore.ListOptions{
+		PageSize: pageSize,
+		Filters:  append(filterForFailureDeploy, commonFilters...),
+	})
+	if err != nil {
+		a.logger.Error("failed to get deployments", zap.Error(err))
+		return nil, status.Error(codes.Internal, "Failed to get deployments")
+	}
+
+	successCount := len(successDeployments)
+	failureCount := len(failureDeployments)
+
+	changeFailureRate := float32(failureCount) / float32(successCount+failureCount)
```

Comment:
The number of deployments can be zero.

Reply:
I still didn't know enough about math...
(I thought `0/0 = 1`)